h2/src/main/org/h2/mvstore/MVStore.java

```diff
@@ -1898,41 +1899,32 @@ public void compactMoveChunks(int targetFillRate, long moveSize) {
         }
     }
 
-    private ArrayList<Chunk> findChunksToMove(long startBlock, long moveSize) {
-        ArrayList<Chunk> move = new ArrayList<>();
-        for (Chunk c : chunks.values()) {
-            if (c.block > startBlock) {
-                move.add(c);
-            }
-        }
-        // sort by block
-        Collections.sort(move, new Comparator<Chunk>() {
-            @Override
-            public int compare(Chunk o1, Chunk o2) {
-                return Long.signum(o1.block - o2.block);
-            }
-        });
-        // find which is the last block to keep
-        int count = 0;
+    private Iterable<Chunk> findChunksToMove(long startBlock, long moveSize) {
+        PriorityQueue<Chunk> queue = new PriorityQueue<>(this.chunks.size() / 2 + 1,
+                new Comparator<Chunk>() {
+                    @Override
+                    public int compare(Chunk o1, Chunk o2) {
+                        return Long.signum(o2.block - o1.block);
```

Comment:
BTW, we can use `Long.compare(long, long)` here now, it is available in all currently supported Java and Android versions.

Code with subtraction raises more attention, because subtraction can overflow (only when arguments may be negative).

Reply:
Agree, I was thinking about replacing signum, but then decided against it because it’s kind of idiomatic by now, and its implementation is more simple (inlinable?).

Reply:
Both methods are inlined (at least with C2 on x86-64), but `signum()` code is branch-free and about 1.5 times faster than `compare()` with random data on my system. Overall difference in performance should be very low. But OK, `block` values cannot be negative here.